### DISCUSSION TITLE
Query option "defaultload" doesn't work with "noload"

### DISCUSSION THREAD

**Discussion by decaz** [sqlalchemy/sqlalchemy#10858] [Usage Questions] [answered]
I have some model with lots of lazy relationships and in different queries I want to disable loading all of them and enable default loading of that I need in specific query.

Here is small example of how `defaultload` option doesn't work in combination with `noload` option:
```python
from sqlalchemy import Column, ForeignKey, Integer, select
from sqlalchemy.orm import (
    declarative_base,
    relationship,
    noload,
    defaultload,
    joinedload,
)

Base = declarative_base()


class A(Base):
    __tablename__ = 'a'

    id = Column(Integer, primary_key=True)


class B(Base):
    __tablename__ = 'b'

    id = Column(Integer, primary_key=True)


class C(Base):
    __tablename__ = 'c'

    id = Column(Integer, primary_key=True)
    a_id = Column(Integer, ForeignKey(A.id), nullable=False)
    b_id = Column(Integer, ForeignKey(B.id), nullable=False)

    a = relationship(A, lazy='joined', innerjoin=True)
    b = relationship(B, lazy='joined', innerjoin=True)


# Load "C" with only "b" joined (don't load "a")
query_1 = select(C).options(noload('*'), defaultload(C.b))
query_2 = select(C).options(noload('*'), joinedload(C.b))

print('query_1:', query_1)
# `defaultload` doesn't work so C.b was not joined
# query_1: SELECT c.id, c.a_id, c.b_id
# FROM c
print('query_2:', query_2)
# `joinedload` works ok so C.b was joined
# query_2: SELECT c.id, c.a_id, c.b_id, b_1.id AS id_1
# FROM c JOIN b AS b_1 ON b_1.id = c.b_id
```

Reproduced in versions 1.4 and 2.0.

**Comment by zzzeek** [comment, marked as answer]
hi -

"noload" has unusual, not-quite-predictable behavior because this loading style doesn't actually change the state of the attribute and it is not generally tested in all the "usual" use cases, because I've yet to see a valid use for "noload" - this option was added many years ago for a particular use case I thought would be useful, but it's not.

After getting a few "noload" issues the past few days I'm 95% sure I want to deprecate and remove this option because it's not easy to support and should not be needed.     Can I ask exactly what it is you need "noload" for that can't be accomplished by simply leaving the attribute at "lazyload()" ?

**Reply by decaz** [reply]
In my case when having a lot of models with lots of lazy relationships and need to join/select/etc. them in different ways for different queries - I think it's better to define all these releationships as lazy "raise" or "raise_on_sql" (maybe it's even better) and continue to use them explicit way without "noload".

Until I have this done I guess "lazyload" would be nice alternative to "noload" loading style.

**Comment by decaz** [comment]
Additional question: can I somehow predefine default loading style for relationship with "raise loading" so I could use "defaultload" everywhere without remembering what loading style I should use for specific relationship? I mean something like that:
```python
b = relationship(B, lazy='raise', lazy_default='join', innerjoin=True)

# Use "joinedload" with inner join as predefined loading style
query_3 = select(C).options(defaultload(C.b))
```

**Reply by zzzeek** [reply]
there are tentative plans to include this in 2.1: #10859 

we've hit law of twos here

**Reply by CaselIT** [reply]
I don't think that was asked here is the same as what was discussed in the linked issue.

@decaz the `lazy=` that is in the relationship declaration *is* the default, so it would be very confusing to have a second default value here.

To do something similar to what you want you can set the wanted loading strategy in the relationship and use in the query
`select(C).options(raiseload('*'), defaultload(C.b))`

**Reply by decaz** [reply]
@CaselIT maybe you are right about two default declarations.. Raise loading exists to prevent unintended access to relationship, but having non-raise loading as a default creates problem that you can unintentionally forget to specify `raiseload('*')` option for some query, so the problem remains.

@zzzeek @CaselIT what do you think about separating raise loading from other loading styles? I think it's not the way (or style) how to lazy load some relationship, - it's restriction for lazy loading. I mean adding some special parameter for "raise loading", something like this:
```python
b = relationship(B, lazy='join', innerjoin=True, lazy_raise=True)
```
or
```python
b = relationship(B, lazy='join', innerjoin=True, lazy_raise='on_sql')
```

**Reply by CaselIT** [reply]
An api like that honestly seems very hard to understand. What does `lazy='join', lazy_raise=True` mean?
that the relationship is loaded with join but would raise if it's not loaded, but when is it not loaded?

**Reply by decaz** [reply]
These are just examples to demonstrate the idea. `lazy_raise` can be anything - `lazy_needs_explicit_option` for instance.

**Reply by CaselIT** [reply]
> lazy_needs_explicit_option 

this seems to be exactly `lazy='raise'` since the end result would be the same: accessing the relationship raises an error.

I'm not following what the objective of this new api would be, could you re-explain what the problem you are trying to solve?

**Reply by decaz** [reply]
@CaselIT you gave an example: `select(C).options(raiseload('*'), defaultload(C.b))`

In case I make a query `select(C)` or `select(C).options(defaultload(C.b))` there will be no raise on loading of any lazy relationships, - that's the problem.

**Reply by CaselIT** [reply]
This seems a particular constraint on your particular code.
I would solve it with a select constructor, something like class method `C.select()` that returns `seleect(cls).options(raiseload('*'))`

**Reply by CaselIT** [reply]
> This seems a particular constraint on your particular code.

by this I mean that I'm not sure how generalizable this requirement is

**Reply by decaz** [reply]
I agree that user code can be used in this case to always add `raiseload('*')` option. My proposition was about to make ORM be able to do the same :)

**Reply by CaselIT** [reply]
@zzzeek do have an opinion on this mike?

**Reply by zzzeek** [reply]
I need to understand the use case better.

is it that in the cases where the attribute gets expired, then it lazy loads later on and that would be preferred as a raise load?  I would tailor a feature towards the "unexpire" use case if that's the case.

otherwise if it's just, "I dont want any relationships to load, but when I do turn them on, I'd like to use something that's configured on the relationship already", there's less convenient ways to roll that yourself, rather than a "default_lazy" I'd want to do something open ended


    relationship(X, lazy="raise", load_profiles={"eager": "joined"})

and then

    defaultload(A.x, profile="eager")

something like that

this is kind of like a CSS class
